esphome/components/mqtt/mqtt_climate.cpp

```diff
@@ -292,35 +294,38 @@ bool MQTTClimateComponent::publish_state_() {
 
   if (traits.get_supports_fan_modes()) {
     const char *payload = "";
-    switch (this->device_->fan_mode.value()) {
-      case CLIMATE_FAN_ON:
-        payload = "on";
-        break;
-      case CLIMATE_FAN_OFF:
-        payload = "off";
-        break;
-      case CLIMATE_FAN_AUTO:
-        payload = "auto";
-        break;
-      case CLIMATE_FAN_LOW:
-        payload = "low";
-        break;
-      case CLIMATE_FAN_MEDIUM:
-        payload = "medium";
-        break;
-      case CLIMATE_FAN_HIGH:
-        payload = "high";
-        break;
-      case CLIMATE_FAN_MIDDLE:
-        payload = "middle";
-        break;
-      case CLIMATE_FAN_FOCUS:
-        payload = "focus";
-        break;
-      case CLIMATE_FAN_DIFFUSE:
-        payload = "diffuse";
-        break;
-    }
+    if (this->device_->fan_mode.has_value())
+      switch (this->device_->fan_mode.value()) {
+        case CLIMATE_FAN_ON:
+          payload = "on";
+          break;
+        case CLIMATE_FAN_OFF:
+          payload = "off";
+          break;
+        case CLIMATE_FAN_AUTO:
+          payload = "auto";
+          break;
+        case CLIMATE_FAN_LOW:
+          payload = "low";
+          break;
+        case CLIMATE_FAN_MEDIUM:
+          payload = "medium";
+          break;
+        case CLIMATE_FAN_HIGH:
+          payload = "high";
+          break;
+        case CLIMATE_FAN_MIDDLE:
+          payload = "middle";
+          break;
+        case CLIMATE_FAN_FOCUS:
+          payload = "focus";
+          break;
+        case CLIMATE_FAN_DIFFUSE:
+          payload = "diffuse";
+          break;
+      }
+    if (this->device_->custom_fan_mode.has_value())
+      payload = this->device_->custom_fan_mode.value().c_str();
```

Comment:
Unfortunately, this is not safe in C++ (if `custom_fan_mode` gets changed during the publish, the `c_str()` no longer points to a valid address).

That can be fixed like this (note the updated payload declaration)

```suggestion
      payload = this->device_->custom_fan_mode.value();
```